Native nftables support for sidecar mode

sridhargaddam opened the pull request:
This PR includes support for using nftables as an alternative to iptables for setting up traffic redirection rules in Istio. 
Nftables offers a more modern and consistent interface, and aligns with the direction many Linux distributions are heading. With this change, users can optionally enable nftables for setting up traffic redirection rules when using the sidecar mode, by configuring the `--native-nftables` flag.

**Supported features:**
* Istio sidecar mode with `istio-cni` as well as `istio-init` container.
* Supports all the usual pod [annotations](https://istio.io/latest/docs/reference/config/annotations/) like excludeInboundPorts, includeOutboundIPRanges, and so on.

**Implementation details:**
* A new config option `--native-nftables` is added which can be set using helm (`global.nativeNftables`) as well as istioctl.
* Updated the injection template and config-map to support this feature.
* The CNI plugin now checks this config and chooses either iptables or nftables accordingly.
* Shared code between iptables and nftables has been moved into a common package to avoid code duplication.
* Added builder APIs that build up nftables rules in memory before applying them to the system.
* Added new APIs that use nftables syntax to support various use cases, including all the existing customization options.
* Added unit tests for the new code paths.
* This implementation uses the [knftables](https://github.com/kubernetes-sigs/knftables) library from the Kubernetes sig-network group. This library is used in couple of other K8s projects like kube-proxy, Calico, OVN-K and so on.

**Why a feature flag?**
This is a new implementation and plays a critical role in traffic redirection, so we want to give it time to mature before making it the default. Introducing a feature flag allows users to adopt it gradually and helps prevent unexpected behavior in environments where both iptables and nftables are available.

**Pending:**
* Support for Ambient mode will be added in a follow-up PR.
* CI jobs for validating the nftables implementation.

**Note:** The default behavior still uses iptables. Nftables will only be used if the feature flag is explicitly enabled.

Related to: https://github.com/istio/istio/issues/47821

Signed-off-by: Sridhar Gaddam <sgaddam@redhat.com>
Co-authored-by: Yuanlin Xu <yuanlin.xu@redhat.com>

sridhargaddam committed:
istio-cni: add support for native nftables rules

This PR supports a new flag `--native-nftables` flag to enable the
use of nftables rules instead of iptables rules for traffic redirection
in Istio.

Signed-off-by: Sridhar Gaddam <sgaddam@redhat.com>
Co-authored-by: Yuanlin Xu <yuanlin.xu@redhat.com>

sridhargaddam committed:
Move common code into a separate package

This PR moves the common code currently in the iptables
package to a separate package, so that it can be shared
between the iptables and nftables implementations without
duplicating code.

Signed-off-by: Sridhar Gaddam <sgaddam@redhat.com>
Co-authored-by: Yuanlin Xu <yuanlin.xu@redhat.com>

sridhargaddam committed:
support native nftables flag while using istio-init container

Signed-off-by: Sridhar Gaddam <sgaddam@redhat.com>
Co-authored-by: Yuanlin Xu <yuanlin.xu@redhat.com>

sridhargaddam committed:
Implement nftables for sidecar proxy mode

This change adds support for using nftables to manage
traffic redirection rules, as an alternative to iptables.
The CNI plugin now uses NftablesInterceptRuleMgr when
enabled, and the istio-iptables tool can set up nftables
rules based on the config. This PR uses knftables library
from sig-network which is a library that is used across
couple of other K8s projects for nftable support.

Signed-off-by: Sridhar Gaddam <sgaddam@redhat.com>
Co-authored-by: Yuanlin Xu <yuanlin.xu@redhat.com>

sridhargaddam committed:
Add unit tests for the nftables code.

Signed-off-by: Sridhar Gaddam <sgaddam@redhat.com>
Co-authored-by: Yuanlin Xu <yuanlin.xu@redhat.com>

sridhargaddam committed:
Add the License file of knftables library

Signed-off-by: Sridhar Gaddam <sgaddam@redhat.com>
Co-authored-by: Yuanlin Xu <yuanlin.xu@redhat.com>

sridhargaddam committed:
Add release notes for nftable support in sidecar mode

Signed-off-by: Sridhar Gaddam <sgaddam@redhat.com>
Co-authored-by: Yuanlin Xu <yuanlin.xu@redhat.com>

istio-testing commented:
@sridhargaddam: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
release-notes_istio | 8eb4aeef9a48fb32d1e1a19f5e9e5d6d2782c101 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/56487/release-notes_istio/1929822579950882816) | true | `/test release-notes`
windowsbuildcheck_istio | 8eb4aeef9a48fb32d1e1a19f5e9e5d6d2782c101 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/56487/windowsbuildcheck_istio/1929822583297937408) | true | `/test windowsbuildcheck`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@sridhargaddam: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
windowsbuildcheck_istio | 2ddf1b3255804997dbc7b3594f7297e49f25f306 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/56487/windowsbuildcheck_istio/1929824583293079552) | true | `/test windowsbuildcheck`
unit-tests_istio | 2ddf1b3255804997dbc7b3594f7297e49f25f306 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/56487/unit-tests_istio/1929824581602775040) | true | `/test unit-tests`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@sridhargaddam: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
windowsbuildcheck_istio | 2ddf1b3255804997dbc7b3594f7297e49f25f306 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/56487/windowsbuildcheck_istio/1929824583293079552) | true | `/test windowsbuildcheck`
unit-tests_istio | 2ddf1b3255804997dbc7b3594f7297e49f25f306 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/56487/unit-tests_istio/1929824581602775040) | true | `/test unit-tests`
unit-tests-arm64_istio | 2ddf1b3255804997dbc7b3594f7297e49f25f306 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/56487/unit-tests-arm64_istio/1929824580768108544) | true | `/test unit-tests-arm64`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

sridhargaddam committed:
Fix unit test failures

Signed-off-by: Sridhar Gaddam <sgaddam@redhat.com>

istio-testing commented:
@sridhargaddam: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
windowsbuildcheck_istio | a477120c0eb4f0957ada742a76af2cd0f736c353 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/56487/windowsbuildcheck_istio/1929851838463479808) | true | `/test windowsbuildcheck`
integ-ambient-mc_istio | a477120c0eb4f0957ada742a76af2cd0f736c353 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/56487/integ-ambient-mc_istio/1929851823091355648) | false | `/test integ-ambient-mc`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

sridhargaddam commented:
Note to reviewers: Sorry for the large PR. I’ve split it into separate commits to make it easier to review. Hope that helps. Thanks!

sridhargaddam committed:
Fix windowsbuildcheck failure

Signed-off-by: Sridhar Gaddam <sgaddam@redhat.com>
Co-authored-by: Yuanlin Xu <yuanlin.xu@redhat.com>

istio-testing commented:
@sridhargaddam: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
integ-ambient-mc_istio | 3f636e272ba75bb9a1333927467c47cc08ce65a3 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/56487/integ-ambient-mc_istio/1930521740161585152) | false | `/test integ-ambient-mc`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

sridhargaddam commented:
> Thank you! I was busy in the last few days. I will take a look tomorrow :)

Thanks @leosarra 🙂

sridhargaddam committed:
Fix DNSRedirect use-case with nftables

This PR addresses two issues.
1. Errors while programming nft rules when using the  DNSRedirect
   use-case
2. It also addresses a duplicate jump rule in the nat table for
   istio-output-dns chain.

Signed-off-by: Sridhar Gaddam <sgaddam@redhat.com>
Co-authored-by: Leonardo Sarra <8293810+leosarra@users.noreply.github.com>

sridhargaddam commented on tools/istio-nftables/pkg/capture/run.go:
```diff
@@ -0,0 +1,1018 @@
+			"udp dport", "53",
+			"meta",
+			"skuid", uid,
+			"CT", "zone", "set", "1")
```
Thanks @leosarra. I was validating the DNSRedirect use-case and noticed this error. The [following commit](https://github.com/istio/istio/pull/56487/commits/4dacac29474fa3df583b5b61bbdda4fd3d4ba3fe) includes this change along with an additional change where a rule was getting duplicated in the nat table with jump target as `istio-output-dns`.

istio-testing commented:
@sridhargaddam: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
integ-ambient-mc_istio | 3f636e272ba75bb9a1333927467c47cc08ce65a3 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/56487/integ-ambient-mc_istio/1930521740161585152) | false | `/test integ-ambient-mc`
gencheck_istio | 4dacac29474fa3df583b5b61bbdda4fd3d4ba3fe | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/56487/gencheck_istio/1932052389183885312) | true | `/test gencheck`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

sridhargaddam committed:
Fix insert rules for raw and mangle tables

Along with supporting the insert operation in the raw and mangle
tables, this PR also optimizes the code to remove duplication.

Signed-off-by: Sridhar Gaddam <sgaddam@redhat.com>
Co-authored-by: Leonardo Sarra <8293810+leosarra@users.noreply.github.com>

istio-testing commented:
@sridhargaddam: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
gencheck_istio | 83b19aa7da50fc224514f310ee2cf78e0eb64b46 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/56487/gencheck_istio/1932069211052445696) | true | `/test gencheck`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

sridhargaddam committed:
Use consistent name for istio-drop chain

This PR renames ISTIO_DROP to istio-drop as well
as fixes an nft rule with "invalid" argument.

Signed-off-by: Sridhar Gaddam <sgaddam@redhat.com>
Co-authored-by: Leonardo Sarra <8293810+leosarra@users.noreply.github.com>

sridhargaddam commented on tools/istio-nftables/pkg/capture/run.go:
```diff
@@ -0,0 +1,1018 @@
+	if dropInvalid {
+		cfg.ruleBuilder.AppendRule(constants.PreroutingChain, constants.IstioProxyMangleTable,
+			"meta l4proto tcp",
+			"ct state", "INVALID",
```
Done, handled in the following commit - https://github.com/istio/istio/pull/56487/commits/792fe6321c96d2d153f29f2688b758dd056ed80b

sridhargaddam committed:
Skip creating empty chains

Signed-off-by: Sridhar Gaddam <sgaddam@redhat.com>
Co-authored-by: Leonardo Sarra <8293810+leosarra@users.noreply.github.com>

sridhargaddam committed:
Support cleanupOnly along with idempotent rules

This PR adds the following support.
1. Support for cleanupOnly flag for VM use-cases.
2. The creation of tables/chains/rules is now idempotent.

Signed-off-by: Sridhar Gaddam <sgaddam@redhat.com>
Co-authored-by: Leonardo Sarra <8293810+leosarra@users.noreply.github.com>

leosarra commented on tools/istio-nftables/pkg/capture/run.go:
```diff
@@ -0,0 +1,1018 @@
+}
+
+// executeCommands runs all the steps to add nftables rules for Istio.
+// It calls separate functions to add rules for NAT, MANGLE, and RAW tables one by one.
```
fyi I have opened a PR for knftables https://github.com/kubernetes-sigs/knftables/pull/24 that provides a batch interface

sridhargaddam committed:
Add e2e for idempotency and cleanup

Signed-off-by: Sridhar Gaddam <sgaddam@redhat.com>
Co-authored-by: Leonardo Sarra <8293810+leosarra@users.noreply.github.com>

istio-testing commented:
@sridhargaddam: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
integ-cni_istio | 4c4889489615ed0b05014a8bcf07793cdb29c258 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/56487/integ-cni_istio/1932506665861517312) | true | `/test integ-cni`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

sridhargaddam commented:
/test integ-cni

sridhargaddam committed:
Ensure that cleanup is properly handled

Signed-off-by: Sridhar Gaddam <sgaddam@redhat.com>
Co-authored-by: Leonardo Sarra <8293810+leosarra@users.noreply.github.com>

dhawton commented:
/retest

sridhargaddam committed:
Log nft rules that are successfully programmed

Signed-off-by: Sridhar Gaddam <sgaddam@redhat.com>

istio-testing commented:
@sridhargaddam: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
unit-tests-arm64_istio | c7ac216ea6ae96f7cecdb99eaafa92829cabbbc5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/56487/unit-tests-arm64_istio/1933569764068691968) | true | `/test unit-tests-arm64`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

sridhargaddam committed:
Use single transaction for all tables

Signed-off-by: Sridhar Gaddam <sgaddam@redhat.com>
Co-authored-by: Leonardo Sarra <8293810+leosarra@users.noreply.github.com>

sridhargaddam commented:
@istio/wg-networking-maintainers can you PTAL, thanks.

sridhargaddam commented:
Hello @howardjohn, @keithmattix /other Istio maintainers,

I'd like to target this Nftables PR for the Istio 1.27 release and would really appreciate your review 🙏🏼. I understand that this is a big PR, so I’ve structured it into individual commits to make it easier to follow. @leosarra has been carefully reviewing the PR,  many thanks to him for the detailed feedback so far. If you have any questions or need clarification on any part, please feel free to ask. Thanks in advance!

leosarra reviewed (approved):
LGTM 👍

istio-testing commented:
PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository.
</details>

keithmattix commented:
@sridhargaddam thank you for this PR and for organizing it. I will get this on my list for review this week

keithmattix commented on tools/istio-nftables/pkg/capture/run.go:
```diff
@@ -0,0 +1,1026 @@
+	HasLoopBackIP bool
+}
+
+func split(s string) []string {
```
Why do you need a separate function just for this?

keithmattix commented on tools/istio-nftables/pkg/capture/run.go:
```diff
@@ -0,0 +1,1026 @@
+	for _, ipRange := range split(cidrList) {
+		ipp, err := netip.ParsePrefix(ipRange)
+		if err != nil {
+			_, err = fmt.Fprintf(os.Stderr, "Ignoring error for bug compatibility with istio-nftables: %s\n", err.Error())
```
Can you give more information for why we ignore this error?

keithmattix commented on tools/istio-nftables/pkg/capture/run.go:
```diff
@@ -0,0 +1,1026 @@
+// The goal here is to make sure incoming traffic reaches Envoy before going to the actual app.
+func (cfg *NftablesConfigurator) handleInboundPortsInclude() {
+	// Handling of inbound ports. Traffic will be redirected to Envoy, which will process and forward
+	// to the local service. If not set, no inbound port will be intercepted by istio nftablesOrFail.
```
What is `nftablesOrFail`?

keithmattix commented on tools/istio-nftables/pkg/capture/run.go:
```diff
@@ -0,0 +1,1026 @@
+	// to the local service. If not set, no inbound port will be intercepted by istio nftablesOrFail.
+	var table string
+	if cfg.cfg.InboundPortsInclude != "" {
+		if cfg.cfg.InboundInterceptionMode == "TPROXY" {
```
Please create a separate function for tproxy for readability

sridhargaddam replied:
Done as part of the following commit - https://github.com/istio/istio/pull/56487/commits/82eee5834024b1e2d186a8cea2c97726a3535d96

keithmattix commented on tools/istio-nftables/pkg/capture/run.go:
```diff
@@ -0,0 +1,1026 @@
+	cfg.shortCircuitExcludeInterfaces()
+	cfg.shortCircuitKubeInternalInterface()
+
+	// Create a rule for invalid drop in PREROUTING chain in mangle table, so the nftables will drop the out of window packets instead of reset connection .
```
I notice that we're still using iptables naming and conventions within nftables; I'm assuming that intentional? Do we think there's a point in time where we may want to migrate to different/simpler tables now that we're leveraging nftables?

keithmattix replied:
Ah just noticed the discussion [here](https://github.com/istio/istio/pull/56487#discussion_r2124607280). Makes sense for a first pass, but I am excited at the prospect of simplifying our redirection logic as much as possible

keithmattix commented on tools/istio-nftables/pkg/capture/run.go:
```diff
@@ -0,0 +1,1026 @@
+	// Create a new chain for redirecting outbound traffic to the common Envoy port.
+	// In both chains, 'counter RETURN' bypasses Envoy and 'counter jump IstioRedirectChain'
+	// redirects to Envoy.
+	cfg.ruleBuilder.AppendRule(
```
I'm an iptables/nftables novice, so for my understanding, could you explain why the outbound redirection to the proxy port is placed before the outbound exclusion logic?

keithmattix commented on tools/istio-nftables/pkg/capture/run.go:
```diff
@@ -0,0 +1,1026 @@
+	// In both chains, 'counter RETURN' bypasses Envoy and 'counter jump IstioRedirectChain'
+	// redirects to Envoy.
+	cfg.ruleBuilder.AppendRule(
+		constants.IstioRedirectChain, constants.IstioProxyNatTable,
```
nit: should this be named IstioOutRedirectChain to match IstioInRedirectChain below?

keithmattix reviewed (commented):
First pass of a review. Mostly just readability nits before I get to the meat of the redirection implementation

yxun commented on tools/istio-nftables/pkg/capture/nftables_api.go:
```diff
@@ -0,0 +1,80 @@
+
+// Dump is part of the interface but not used in the real implementation. It's used as part of unit tests.
+func (r *RealNftables) Dump(tx *knftables.Transaction) string {
+	// We do not use Dump in the real Interface.
```
Initially, we used the Dump method for comparing results in unit tests. However, it changes the order of chains and rules alphabetically. Then we decided not to use this method and found a Transaction.String method instead. 
yes, we will remove this Dump method in the interface and the implementation.

sridhargaddam commented on go.mod:
```diff
@@ -2,6 +2,8 @@ module istio.io/istio
 
 go 1.24.0
 
+replace sigs.k8s.io/knftables => github.com/danwinship/knftables v0.0.17-0.20250612183636-0ab847ae97fe
```
I spoke with Dan Winship (one of the maintainers of knftables), and he mentioned that he’ll be making a point release soon. I’ll update the PR to use the new release once it’s available.

sridhargaddam commented on tools/istio-nftables/pkg/capture/run.go:
```diff
@@ -0,0 +1,1026 @@
+}
+
+// NetworkRange is used to hold IPv4 or IPv6 ranges.
+type NetworkRange struct {
```
Good catch. Its used only within the same package, I'll update it.

sridhargaddam committed:
Refactor and address review comments

Signed-off-by: Sridhar Gaddam <sgaddam@redhat.com>

sridhargaddam committed:
Create separate function for TPROXY handling

Signed-off-by: Sridhar Gaddam <sgaddam@redhat.com>

sridhargaddam committed:
Fix unit test failures in cni/install pkg

Signed-off-by: Sridhar Gaddam <sgaddam@redhat.com>

keithmattix reviewed (commented):
A few nits, but otherwise, this LGTM! Might be worth adding an integration test, but that can be a follow up IMO

sridhargaddam commented on tools/istio-iptables/pkg/cmd/root.go:
```diff
@@ -213,7 +226,7 @@ func ProgramIptables(cfg *config.Config) error {
 			return err
 		}
-		if err := capture.ConfigureRoutes(cfg); err != nil {
+		if err := tproxy.ConfigureRoutes(cfg); err != nil {
```
We refactored existing code into a common tproxy pkg so it can be used by both iptables and nftables impl. Inside the ConfigureRoutes API, we still have check for cfg.InboundInterceptionMode==TPROXY.

sridhargaddam committed:
Use latest knftables as the PR is now merged

Signed-off-by: Sridhar Gaddam <sgaddam@redhat.com>

sridhargaddam reviewed (commented):
Thanks for reviewing @keithmattix.

keithmattix reviewed (approved):
I'm OOO next week, so approving for networking with a hold so that @howardjohn can take a look before feature freeze. If there's no feedback by then, remove the hold

istio-testing commented:
PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository.
</details>

sridhargaddam committed:
Rebase and address merge conflicts

Signed-off-by: Sridhar Gaddam <sgaddam@redhat.com>

istio-testing commented:
@sridhargaddam: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
unit-tests-arm64_istio | 65974705eb57eb7886dad7e1c3302ea0fdc19f2d | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/56487/unit-tests-arm64_istio/1943559086469550080) | true | `/test unit-tests-arm64`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@sridhargaddam: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
unit-tests-arm64_istio | 65974705eb57eb7886dad7e1c3302ea0fdc19f2d | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/56487/unit-tests-arm64_istio/1943559086469550080) | true | `/test unit-tests-arm64`
unit-tests_istio | 65974705eb57eb7886dad7e1c3302ea0fdc19f2d | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/56487/unit-tests_istio/1943559087291633664) | true | `/test unit-tests`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

sridhargaddam commented:
> I'm OOO next week, so approving for networking with a hold so that @howardjohn can take a look before feature freeze. If there's no feedback by then, remove the hold

@zirain @keithmattix I removed the hold, but the PR is not getting merged. Any idea why?

sridhargaddam commented:
> > @zirain @keithmattix I removed the hold, but the PR is not getting merged. Any idea why?
> 
> Because it's missing reviews from one of the workgroups @howardjohn is part of. He commented so it's not listed, but `tide` says you need an approval from more than Keith.

Thanks for the clarification @dhawton. It seems like @howardjohn is a bit tied up at the moment. With the feature freeze coming up next week, I'm a bit concerned about the timelines. Would it be possible for someone else from the relevant workgroup to take a look and approve the PR in the meantime?

dgn commented:
ping @istio/technical-oversight-committee - PTAL, we would like to merge this before the feature freeze if possible

sridhargaddam commented:
/test gencheck

istio-testing commented:
@sridhargaddam: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
gencheck_istio | 65974705eb57eb7886dad7e1c3302ea0fdc19f2d | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/56487/gencheck_istio/1944723349749895168) | true | `/test gencheck`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

sridhargaddam commented:
/test gencheck

sridhargaddam commented:
The failure in "gencheck" is not related to this PR.

istio-testing commented:
In response to a cherrypick label: new issue created for failed cherrypick: #56996
